zcash_client_backend/src/fees/zip317.rs

```diff
@@ -67,129 +71,157 @@ impl ChangeStrategy for SingleOutputChangeStrategy {
-            // order of the original input slices, with the most preferred inputs first.
-            transparent_dust.reverse();
-            transparent_dust.truncate(t_disallowed_dust);
-            sapling_dust.reverse();
-            sapling_dust.truncate(s_disallowed_dust);
-            orchard_dust.reverse();
-            orchard_dust.truncate(o_disallowed_dust);
+            let mut orchard_dust: Vec<NoteRefT> = vec![];
 
+            // Depending on the shape of the transaction, we may be able to spend up to
+            // `grace_actions - 1` dust inputs. If we don't have any dust inputs though,
+            // we don't need to worry about any of that.
             if !(transparent_dust.is_empty() && sapling_dust.is_empty() && orchard_dust.is_empty())
             {
-                return Err(ChangeError::DustInputs {
-                    transparent: transparent_dust,
-                    sapling: sapling_dust,
+                let t_non_dust = transparent_inputs.len() - transparent_dust.len();
+                let t_allowed_dust = transparent_outputs.len().saturating_sub(t_non_dust);
+
+                // We add one to either the Sapling or Orchard outputs for the (single)
+                // change output. Note that this means that wallet-internal shielding
+                // transactions are an opportunity to spend a dust note.
+                let net_flows = calculate_net_flows::<NoteRefT, Self::FeeRule, Self::Error>(
+                    transparent_inputs,
+                    transparent_outputs,
+                    sapling,
                     #[cfg(feature = "orchard")]
-                    orchard: orchard_dust,
-                });
+                    orchard,
+                    #[cfg(feature = "transparent-inputs")]
+                    ephemeral_parameters.ephemeral_input_amount(),
+                    #[cfg(feature = "transparent-inputs")]
+                    ephemeral_parameters.ephemeral_output_amount(),
+                )?;
+                let (_, sapling_change, orchard_change) =
+                    single_change_output_policy(&net_flows, self.fallback_change_pool);
+
+                let s_non_dust = sapling.inputs().len() - sapling_dust.len();
+                let s_allowed_dust =
+                    (sapling.outputs().len() + sapling_change).saturating_sub(s_non_dust);
+
+                #[cfg(feature = "orchard")]
+                let (orchard_inputs_len, orchard_outputs_len) =
+                    (orchard.inputs().len(), orchard.outputs().len());
+                #[cfg(not(feature = "orchard"))]
+                let (orchard_inputs_len, orchard_outputs_len) = (0, 0);
+
+                let o_non_dust = orchard_inputs_len - orchard_dust.len();
+                let o_allowed_dust =
+                    (orchard_outputs_len + orchard_change).saturating_sub(o_non_dust);
+
+                let available_grace_inputs = self
```

Comment:
This calculation is wrong. We want to calculate how many additional inputs can be spent as dust without increasing the fee as a result of the grace actions (*not* matching input slots with existing outputs). Since `grace_actions()` is 2 and it cannot happen that the number of actions is 0, there is at most one slot to spend another shielded input for free up to the grace action threshold. This occurs only when the number of actions was otherwise 1, i.e. when `max(t_non_dust, transparent_outputs.len()) + max(s_non_dust, sapling_outputs.len()) + max(o_non_dust, orchard_outputs.len()) == 1`. If we don't take into account outputs, then we will incorrectly double-count opportunities for spending dust inputs that were already counted in `{t,s,o}_allowed_dust`.

Reply:
Fixed.